common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java

```diff
@@ -140,4 +166,40 @@ private String getCurrentActiveBrokerPackageName(@NonNull final Context context)
 
         return null;
     }
+
+    /**
+     * Returns the inactive installed authenticator.
+     * - If CP is the broker, this will return authenticator.
+     * - If authenticator is the broker, this will return CP.
+     *
+     * @param context
+     * @return package name of the installed inactive broker app, if there's any.
+     */
+    @Nullable
+    private String getInactiveBrokerPackageName(@NonNull final Context context) {
+        final String methodName = ":getInactiveBrokerPackageName";
+        String currentBroker = getCurrentActiveBrokerPackageName(context);
+
+        final String inactiveBrokerPackageName;
+        if (AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(currentBroker)) {
+            inactiveBrokerPackageName = AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME;
+        } else if (AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(currentBroker)) {
+            inactiveBrokerPackageName = AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME;
+        } else {
+            return null;
+        }
+
+        // Verify the signature to make sure that we're not binding to malicious apps.
+        final BrokerValidator validator = new BrokerValidator(context);
+        try {
+            if (validator.verifySignature(inactiveBrokerPackageName)) {
+                return inactiveBrokerPackageName;
+            }
```

Comment:
May be we should log here if verification fails

Reply:
Yes, actually, we should send telemetry if the signature check fails.

---
In reply to: [299733100](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/538#discussion_r299733100) [](ancestors = 299733100)